### ISSUE TITLE
Polish and improve the index selection file policy 

### ISSUE THREAD

**Issue by lee-lei** [Intel-bigdata/OAP#509]
In our index scanners, the index selection file policy is to compare if the index file size is greater than the ratio(default is 0.7) of oap data file size. If greater, then use full scan. If not, use index. However, the oap data file size here is the compressed size rather than the raw uncompressed file size. This comparison is unfair. The "SNAPPY" and "GZIP" have the good compression ratio. Thus the compressed oap data file size is much smaller. With this policy, it will use full scan rather than index. This is unexpected. So far, this file policy is disabled by default for all the unit tests.

The improvement will get the fair comparison between uncompressed index file size and uncompressed oap data file size. And it will be done in the next release(0.4.0).

**Comment by chenghao-intel** [comment]
I don't think this is valid, as we need to check the filtered record ratio, other than the index file size.

**Comment by lee-lei** [comment]
I am not sure what do you mean by filtered record ratio. What I meant is the code in lines 121 - 126 of IndexScanner.scala. IMHO, if it's invalid, it should be removed.

**Comment by chenghao-intel** [comment]
@yueawang @carsonwang  can you please check the index selection strategy for BitMap? Sorry per my understanding, the index selection will reply on the filtered out record count (ratio), not the index file size itself. Anything not follow our design?

**Comment by yueawang** [comment]
There are 2 policies (file size + statistics) which are applied to query together. The file size policy plays a IO cost role, however, we should have better prediction of file io cost.
So far I think roaring bitmap file size is impossible ~= db file size, so this policy should not take effect here.

**Comment by lee-lei** [comment]
The key issue is that in this file size policy, the oap data file size is compressed with gizp format. The index file size is not compressed. It may lead to incorrect result (expected to use index may cause unexpected full scan) to compare the uncompressed index file size with the compressed oap data file size. This will impact both btree and bitmap index, from my point of view.

**Comment by lee-lei** [comment]
By default the gzip appends the uncompressed size in the tail 4 byte of the file. However, the snappy and lz4 doesn't have it. For the solution of this issue, I plan to append the raw uncompressed size in the tail 4 byte of oap data file size. This will keep backward compatibility. Meanwhile it's convenient and fast to get the uncompressed file size. This issue doesn't impact the functionality of OAP. The plan is to fix it in 0.4.0 release.

**Comment by chenghao-intel** [comment]
I think we need a unified standard to judge the index size, not only for BitMap Index, but also should apply to B+Tree Index.

**Comment by chenghao-intel** [comment]
And maybe we have to update this to 0.3 as a bug fixing for index statistic. please focus on the existing issues related this problem like #511 .

**Comment by lee-lei** [comment]
I would remind that this issue is not related to bitmap index or btree index. The index file (both bitmap and btree) are not compressed with any codec (gzip, snappy, lz4, etc). The oap data file is compressed with gzip by default. So the fair approach is to compared the uncompressed file size between index and oap data. The roaring bitmap is using run-length encoding to create bitmap of a list of integers. They are not related with the file size policy from my point of view.

**Comment by lee-lei** [comment]
For OAP file, we can use 8 bytes to append the total uncompressed size in each OAP file. Thus it's easy to get it to compare with index file size for index scanner. For Parquet file, it may not be appropriate to modify the default Parquet format to append the raw uncopressed size. One empirical approach is to consider the average snappy compression ratio before comparing the parquet file size. Any comments? Thanks.

**Comment by lee-lei** [comment]
The #543 is the solution for 0.3.0 release. In the next release, we can refine this policy.
